### ISSUE TITLE
k8sm scheduler should ignore scheduling static/mirror pods 

### ISSUE THREAD

**Issue by ravilr** [mesosphere/kubernetes-mesos#493]
@jdef 
using k8sm-v1.0.5-0.6.4 release.
we are seeing the k8sm scheduler is trying to handle static pods (kubelet pod source : file). it should ignore handling pods with  annotations[kubelet.ConfigMirrorAnnotationKey] set to 'mirror'.

- During initial recovery,  :

I0910 22:54:28.816690    3037 scheduler.go:220] initializing kubernetes mesos scheduler
E0910 22:54:28.830373    3037 scheduler.go:927] failed to recover task from pod, will attempt to delete 'kube-system/controller-manager-master-yyy': incomplete metadata: missing value for pod annotation: k8s.mesosphere.io/bindingHost
E0910 22:54:29.386419    3037 scheduler.go:927] failed to recover task from pod, will attempt to delete 'kube-system/kube-controller-manager-yyy': incomplete metadata: missing value for pod annotation: k8s.mesosphere.io/bindingHost

- During regular scheduler loop:

I0913 21:14:24.116840    3037 plugin.go:250] Try to schedule pod controller-manager-master-yyy
I0913 21:14:24.116913    3037 fcfs.go:89] failed to find a fit for pod: kube-system/controller-manager-master-yyy
I0913 21:14:24.116920    3037 plugin.go:746] Failed to schedule: &{{ } {controller-manager-master-yyy  kube-system /api/v1/namespaces/kube-system/pods/controller-manager-master-yyy e48fea03-580e-11e5-8e6e-d4ae528e30a0 657493 0 2015-09-10 22:54:29 +0000 UTC <nil> map[] map[kubernetes.io/config.seen:2015-09-10T22:01:43.528148404Z kubernetes.io/config.source:file kubernetes.io/config.mirror:mirror]} {[{k8s {0xc2089a99d0 <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil>}} {manifests {0xc2089a99f0 <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil> <nil>}}] [{controller-manager-elector gcr.io/google_containers/podmaster:1.1 [/podmaster --etcd-servers=http://y.y.y.y:2379,http://y.y.y.y:2379,http://y.y.y.y:2379 --key=controller --source-file=/kubernetes/kube-controller-manager-v1.yaml --dest-file=/manifests/kube-controller-manager-v1.yaml --ttl-secs=120 --sleep=30s] []  [] [] {map[] map[]} [{k8s true /kubernetes} {manifests false /manifests}] <nil> <nil> <nil> /dev/termination-log IfNotPresent <nil>}] Always <nil> <nil> ClusterFirst map[]  hostname.com true []} {Running [{Ready True}]   206.190.37.225 206.190.37.225 2015-09-10 22:54:29 +0000 UTC [{controller-manager-elector {<nil> 0xc2089dc160 <nil>} {<nil> <nil> 0xc2083cfc70} true 1 gcr.io/google_containers/podmaster:1.1 docker://0ef0696dc670e21cf403e92287e6ab664255a26840e9d0abd5286fde4cca8194 docker://b9c45f86b629521b98a28369d7cc38ba50d28507c9e4f6d3c2a2208364e0a436}]}}
I0913 21:14:24.117094    3037 plugin.go:513] Error scheduling controller-manager-master-yyy: No suitable offers for pod/task; retrying

**Comment by jdef** [comment]
Thanks for reporting this. Will take a look

**Comment by jdef** [comment]
xref https://github.com/kubernetes/kubernetes/pull/13918

**Comment by jdef** [comment]
xref https://github.com/kubernetes/kubernetes/pull/13921 : support for custom annotations w/ static pod sources

**Comment by jdef** [comment]
static pods on mesos/docker cluster are not stable:
```
I1020 20:48:58.492697     196 manager.go:1705] Found pod infra container for "busybox-172.17.0.105_default"
I1020 20:48:58.493221     196 manager.go:1718] Pod infra container looks good, keep it "busybox-172.17.0.105_default"
I1020 20:48:58.493280     196 manager.go:1741] pod "busybox-172.17.0.105_default" container "busybox" exists as d37d5ba881b4304a95a17160368124d7d4a53558898cb1edd13ccb5fa70940d5
I1020 20:48:58.493298     196 manager.go:1819] Got container changes for pod "busybox-172.17.0.105_default": {StartInfraContainer:false InfraContainerId:2741f9284bd043de3d9566100bbf94d0f10d1c289a363d05d84ae67bcce1287a ContainersToStart:map[] ContainersToKeep:map[2741f9284bd043de3d9566100bbf94d0f10d1c289a363d05d84ae67bcce1287a:-1 d37d5ba881b4304a95a17160368124d7d4a53558898cb1edd13ccb5fa70940d5:0]}
I1020 20:48:58.493320     196 kubelet.go:1419] Creating a mirror pod "busybox-172.17.0.105_default"
E1020 20:48:58.497157     196 kubelet.go:1421] Failed creating a mirror pod "busybox-172.17.0.105_default": pods "busybox-172.17.0.105" already exists
```

**Comment by jdef** [comment]
xref https://github.com/kubernetes/kubernetes/issues/15960

**Comment by jdef** [comment]
the error i'm seeing are most likely due to the fact that our kubelet-executor isn't watching the apiserver and so it isn't seeing the mirror pods created there. xref https://github.com/kubernetes/kubernetes/issues/6651

**Comment by jdef** [comment]
fixes in place. blocked by changes for #619; has undergone two rounds of preliminary review.

**Comment by ravilr** [comment]
@sttts can this be cherry-pick'ed to v0.7.0-v1.1.1 please. Thanks.

**Comment by sttts** [comment]
@ravilr done

**Comment by ravilr** [comment]
@jdef 
running v0.7.2-1.1.5 release, still seeing scheduler trying to schedule a static pod. in my case, it is a static pod launched for the controller-manager component on the master k8s host (not on mesos slaves). we run a k8s stock kubelet to launch controller-manager on master host using podmaster pod for HA.  

k8sm scheduler keeps spewing these messages every loop cycle:

I0203 20:09:39.080245    6249 algorithm.go:79] Try to schedule pod kube-controller-manager-m1.www.com
I0203 20:09:39.080285    6249 fcfs.go:95] failed to find a fit for pod: kube-system/kube-controller-manager-m1.www.com
I0203 20:09:39.080289    6249 controller.go:88] Failed to schedule: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:kube-controller-manager-m1.www.com GenerateName: Namespace:kube-system SelfLink:/api/v1/namespaces/kube-system/pods/kube-controller-manager-m1.www.com UID:6522d283-ca21-11e5-a652-fa163eee7610 ResourceVersion:261360 Generation:0 CreationTimestamp:2016-02-03 02:54:08 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[**kubernetes.io/config.mirror:fbe37493e2bdc5b48b67a0f9222e6ab3 **kubernetes.io/config.seen:2016-02-03T02:53:38.216301406Z kubernetes.io/config.source:file kubernetes.io/config.hash:fbe37493e2bdc5b48b67a0f9222e6ab3]} Spec:{Volumes:[{Name:kubecloudconfig VolumeSource:{HostPath:0xc2088f4a20 EmptyDir:<nil> GCEPersistentDisk:<nil> AWSElasticBlockStore:<nil> GitRepo:<nil> Secret:<nil> NFS:<nil> ISCSI:<nil> Glusterfs:<nil> PersistentVolumeClaim:<nil> RBD:<nil> Cinder:<nil> CephFS:<nil> Flocker:<nil> DownwardAPI:<nil> FC:<nil>}} {Name:logdir VolumeSource:{HostPath:0xc2088f4a40 EmptyDir:<nil> GCEPersistentDisk:<nil> AWSElasticBlockStore:<nil> GitRepo:<nil> Secret:<nil> NFS:<nil> ISCSI:<nil> Glusterfs:<nil> PersistentVolumeClaim:<nil> RBD:<nil> Cinder:<nil> CephFS:<nil> Flocker:<nil> DownwardAPI:<nil> FC:<nil>}} {Name:etcssl VolumeSource:{HostPath:0xc2088f4a60 EmptyDir:<nil> GCEPersistentDisk:<nil> AWSElasticBlockStore:<nil> GitRepo:<nil> Secret:<nil> NFS:<nil> ISCSI:<nil> Glusterfs:<nil> PersistentVolumeClaim:<nil> RBD:<nil> Cinder:<nil> CephFS:<nil> Flocker:<nil> DownwardAPI:<nil> FC:<nil>}} {Name:etcpkitls VolumeSource:{HostPath:0xc2088f4a80 EmptyDir:<nil> GCEPersistentDisk:<nil> AWSElasticBlockStore:<nil> GitRepo:<nil> Secret:<nil> NFS:<nil> ISCSI:<nil> Glusterfs:<nil> PersistentVolumeClaim:<nil> RBD:<nil> Cinder:<nil> CephFS:<nil> Flocker:<nil> DownwardAPI:<nil> FC:<nil>}}] Containers:[{Name:kube-controller-manager Image:docker-registry.com:4443/s/k8sm:v1.1.5_1 Command:[/bin/sh -c /usr/bin/km controller-manager --master=1.1.1.1:8080 --cloud-provider=mesos --cloud-config=/etc/k8sm/mesos-cloud --host_port_endpoints=true --v=2 --log_dir=/var/log/k8sm/controller-manager --pod-eviction-timeout=3m40s --deleting-pods-qps=0.02 --logtostderr=false] Args:[] WorkingDir: Ports:[] Env:[{Name:GLOG_max_log_size Value:200 ValueFrom:<nil>}] Resources:{Limits:map[cpu:{Amount:0.250 Format:DecimalSI} memory:{Amount:67108864 Format:BinarySI}] Requests:map[cpu:{Amount:0.250 Format:DecimalSI} memory:{Amount:67108864 Format:BinarySI}]} VolumeMounts:[{Name:kubecloudconfig ReadOnly:true MountPath:/etc/k8sm/mesos-cloud} {Name:logdir ReadOnly:false MountPath:/var/log/k8sm/controller-manager} {Name:etcssl ReadOnly:true MountPath:/etc/ssl} {Name:etcpkitls ReadOnly:true MountPath:/etc/pki/tls}] LivenessProbe:0xc2088fe3f0 ReadinessProbe:<nil> Lifecycle:<nil> TerminationMessagePath:/dev/termination-log ImagePullPolicy:IfNotPresent SecurityContext:<nil> Stdin:false StdinOnce:false TTY:false}] RestartPolicy:Always TerminationGracePeriodSeconds:0xc2088f4b38 ActiveDeadlineSeconds:<nil> DNSPolicy:ClusterFirst NodeSelector:map[] ServiceAccountName: NodeName:m1.www.com HostNetwork:true HostPID:false HostIPC:false ImagePullSecrets:[]} Status:{Phase:Running Conditions:[{Type:Ready Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:0001-01-01 00:00:00 +0000 UTC Reason: Message:}] Message: Reason: HostIP:1.1.1.1 PodIP:1.1.1.1 StartTime:2016-02-03 02:54:08 +0000 UTC ContainerStatuses:[{Name:kube-controller-manager State:{Waiting:<nil> Running:0xc2088fb340 Terminated:<nil>} LastTerminationState:{Waiting:<nil> Running:<nil> Terminated:<nil>} Ready:true RestartCount:0 Image:docker-registry.com:4443/s/k8sm:v1.1.5_1 ImageID:docker://d37577cb758a0360b114d359430385e11514ceeba36a90ce303a187e6e5500d1 ContainerID:docker://b13451f266eb50910908d522a8a3a8da5d58ab9c5a1d1d4d1fb6269ff857a505}]}}
I0203 20:09:39.080546    6249 errorhandler.go:59] Error scheduling kube-controller-manager-m1.www.com: No suitable offers for pod/task; retryin

**Comment by jdef** [comment]
Likely happening because you're using the stock kubelet, which doesn't
apply k8sm annotations that our scheduler expects. I suspect that if you
run the stock kubelet in "runonce" mode (which should disconnect it from
the apiserver) then this problem would go away. The desire to mix stock
kubelet and kubelet-executor is one that I haven't considered before, but
seems valid. If it's an important capability for you (sounds like it is)
can you open a new ticket to track it (and reference this one)? Thanks!

**Comment by jdef** [comment]
Also, could you please elaborate on your use case?

For example, there's been some discussion about using kubelets like this to
*bootstrap* a cluster, but once a base cluster is up and running to use a
deployment (or RC) to bring up redundant components and then shut down the
bootstrap ones. In this case that would mean bringing up a redundant
controller-manager via RC (or deployment), that's leader election aware,
and then shutting down the bootstrap kubelet. Is that what you're trying to
achieve?

**Comment by ravilr** [comment]
Currently, we are running stock kubelet on 3 redundant k8sm master hosts and having it launch a pod-master pod on all 3 hosts for lease locking and master election through etcd of controller-manager's to run on one of those 3 hosts. this is as per the doc here: https://github.com/kubernetes/kubernetes/blob/master/docs/admin/high-availability.md#master-elected-components

Looks like the lease locking logic is retrofitted into controller-manager binary itself now in upstream k8s :
https://github.com/kubernetes/kubernetes/pull/19621
which makes the above pod-master setup obsolete. we will move away from this setup once we land on the newer k8s-1.2 version.

I will check if the runonce mode of kubelet works with the pod-master setup and use it in the time being. Since, the setup expects the kubelet to watch the manifest dir and start/stop controller-manager daemon on one of the hosts, runonce mode may not work in this case.
